### PR TITLE
Aj/sload

### PR DESCRIPTION
**Pull request by veriAjinkya**
1. Load type of sload is now inferred from pointer instead of assumed to be fixed
2. In json object describing the struct type, key field now referes to array of members (previously it was a nested object, with each field name as key and its description as a object)
        This is required for traversing the fields in order they appear, else the sequence of fields gets screwed up and therefore distances
3. Fixed distance calculation for byte offset to field index.

### CODE REVIEW

File: src/pyul/preprocess.py [hunk 1]
```diff
@@ -227,15 +227,15 @@ def parse_type(type_dict: dict) -> ast.YulType:
...
+                    fields=[
+                        ast.YulStateVar(
                             name=td["label"],
                             type=td["type"],
                             slot=int(td["slot"]),
                             offset=int(td["offset"]),
                         )
                         for td in type_dict["members"]
-                    },
+                    ],
```
Comment (Technius) [inline src/pyul/preprocess.py:238]: Please also update the part of the readme that describes the metadata format.

**Review by Technius — COMMENTED** [review]
Waiting for #101 to be merged before I review this further.

File: lib/include/libYulAST/YulConstants.h [hunk 2]
```diff
@@ -224,4 +226,6 @@ const static std::string READ_FROM_MEMORY_REGEX_LIT =
     R"(^read_from_memory(.*?)$)";
 const static std::string WRITE_TO_MEMORY_REGEX_LIT =
     R"(^write_to_memory_(.*?)$)";
+const static std::string COPY_STRUCT_TO_STOTRAGE_REGEX_LIT =
```
Comment (Technius) [inline lib/include/libYulAST/YulConstants.h:229]:
```suggestion
const static std::string COPY_STRUCT_TO_STORAGE_REGEX_LIT =
```

File: lib/libyul2llvm/YulASTVisitor/IntrinsicEmitter.cpp [hunk 3]
```diff
@@ -546,3 +555,26 @@ YulIntrinsicHelper::handleMLoadFunctionCall(YulFunctionCallNode &node) {
...
+  CopyStructToStorageParseResult res =
+      patternMatcher.parseCopyStructToStorage(node.getCalleeName().data());
+  llvm::Value *srcPtr = nullptr, *destPtr = nullptr;
+  destPtr = visitor.visit(*(node.getArgs()[0]));
+  srcPtr = visitor.visit(*(node.getArgs()[1]));
+  if (!destPtr->getType()->isPointerTy() ||
+      destPtr->getType()->getPointerAddressSpace() !=
+          static_cast<int>(YulAddressSpace::STORAGE) ||
+      !destPtr->getType()->getPointerElementType()->isStructTy()) {
+    llvm::errs() << "Source pointer not a pointer to storage struct";
```
Comment (Technius) [inline lib/libyul2llvm/YulASTVisitor/IntrinsicEmitter.cpp:571]: This seems to check the dest pointer. The source pointer is not checked.

**Review by Technius — CHANGES_REQUESTED** [review]
This needs a test case that checks the logic in this PR.

File: lib/include/libYulAST/YulConstants.h (hunk 2 above)
Comment (veriAjinkya) [reply lib/include/libYulAST/YulConstants.h:229]: done

File: lib/libyul2llvm/YulASTVisitor/IntrinsicEmitter.cpp (hunk 3 above)
Comment (veriAjinkya) [reply lib/libyul2llvm/YulASTVisitor/IntrinsicEmitter.cpp:571]: added checks

File: src/pyul/preprocess.py (hunk 1 above)
Comment (veriAjinkya) [reply src/pyul/preprocess.py:238]: done

**Review by Technius — APPROVED** [review]
LGTM contingent on a test case for this PR

**Comment by veriAjinkya** [comment]
Tkt created for the PR so we dont miss out on it

**Pull request merged**